src/main/java/org/checkerframework/checker/crypto/CryptoAnnotatedTypeFactory.java

```diff
@@ -1,57 +1,123 @@
+         */
+        public CryptoQualifierHierarchy(
+                Collection<Class<? extends Annotation>> qualifierClasses, Elements elements) {
+            super(qualifierClasses, elements);
+            ALLOWEDALGORITHMS_KIND = getQualifierKind(ALLOWEDALGORITHMS);
+            ALLOWPROVIDERS_KIND = getQualifierKind(ALLOWPROVIDERS);
         }
 
         @Override
-        public boolean isSubtype(final AnnotationMirror subtype, final AnnotationMirror supertype) {
-            if (AnnotationUtils.areSameByClass(supertype, UnknownAlgorithmOrProvider.class)
-                    || AnnotationUtils.areSameByClass(subtype, Bottom.class)) {
+        public boolean isSubtypeWithElements(
+                AnnotationMirror subAnno,
+                QualifierKind subKind,
+                AnnotationMirror superAnno,
+                QualifierKind superKind) {
+            /* old */
+            if (AnnotationUtils.areSameByClass(superAnno, UnknownAlgorithmOrProvider.class)
+                    || AnnotationUtils.areSameByClass(subAnno, Bottom.class)) {
                 return true;
-            } else if (AnnotationUtils.areSameByClass(subtype, UnknownAlgorithmOrProvider.class)
-                    || AnnotationUtils.areSameByClass(supertype, Bottom.class)) {
+            } else if (AnnotationUtils.areSameByClass(subAnno, UnknownAlgorithmOrProvider.class)
+                    || AnnotationUtils.areSameByClass(superAnno, Bottom.class)) {
                 return false;
-            } else if (AnnotationUtils.areSameByClass(subtype, AllowedAlgorithms.class)
-                    && AnnotationUtils.areSameByClass(supertype, AllowedAlgorithms.class)) {
-                return compareAllowedAlgorithmOrProviderTypes(subtype, supertype);
-            } else if (AnnotationUtils.areSameByClass(subtype, AllowedProviders.class)
-                    && AnnotationUtils.areSameByClass(supertype, AllowedProviders.class)) {
-                return compareAllowedAlgorithmOrProviderTypes(subtype, supertype);
+            } else if (AnnotationUtils.areSameByClass(subAnno, AllowedAlgorithms.class)
+                    && AnnotationUtils.areSameByClass(superAnno, AllowedAlgorithms.class)) {
+                return compareAllowedAlgorithmOrProviderTypes(subAnno, superAnno);
+            } else if (AnnotationUtils.areSameByClass(subAnno, AllowedProviders.class)
+                    && AnnotationUtils.areSameByClass(superAnno, AllowedProviders.class)) {
+                return compareAllowedAlgorithmOrProviderTypes(subAnno, superAnno);
             } else {
                 return false;
             }
         }
 
+        @Override
+        protected AnnotationMirror greatestLowerBoundWithElements(
+                AnnotationMirror a1,
+                QualifierKind qualifierKind1,
+                AnnotationMirror a2,
+                QualifierKind qualifierKind2,
+                QualifierKind glbKind) {
+            if (qualifierKind1.isBottom()) {
```

Comment:
Doesn't this logic always return bottom?

Reply:
Yeah, I checked it, always returning BOTTOM is fine.